enhancements/separating-provider-specific-code-in-the-installer.md

```diff
@@ -0,0 +1,147 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+This change introduces an internal API in the installer code to move provider-specific code into a separate folder and avoid having large switch-case structures in the installer code.
+
+## Motivation
+
+Several parts of the installer code have large, [unwieldy switch-case structures](https://github.com/openshift/installer/blob/254680f6268a7b8e57091d788256c67b81755b5d/pkg/asset/cluster/tfvars.go#L166) that contain provider-specific codes. This makes it hard to keep track of the code that serves a provider, and it makes it impossible to write component-level tests. Creating a separated API reduces friction between teams and makes changes easier to merge because the risk of accidental side effects between providers is reduced. 
+
+### Goals
+
+Create a Go interface and the hooks in the installer code that separatees the installer core code from the provider-specific code.
+
+### Non-Goals
+
+Remove all provider-specific pieces from the code. That would require a complete refactor of the install config data structure and move the Terraform code. 
+
+## Proposal
+
+### Step 1: Create a Go interface
+
+We should create a Go interface that describes the functionality each provider must implement. For example, from the prototype:
+
+```go
+type PlatformV2 interface {
+	// Name returns the name of the platform.
+	Name() string
+
+	// Metadata translates the installConfig to the appropriate entry in clusterMetadata.
+	Metadata(clusterMetadata *types.ClusterMetadata, installConfig *installconfig.InstallConfig) error
+
+    //...
+}
+```
+
+Not all platforms implement all features. Optional features should be hidden behind a feature flag and a separate interface:
+
+```go
+var ErrNotAnIPIPlatform = errors.New("...")
+
+type PlatformV2 interface {
+    //...
+
+    // GetIPI returns the IPI-specific interface, or ErrNotAnIPIPlatform if the platform does not support IPI.
+	GetIPI() (IPIPlatformV2, error)
+}
+
+type IPIPlatformV2 interface {
+    // IPI-specific 
+}
+``` 
+
+### Step 2: Implement registry
+
+In this step we should implement a registry pattern where providers can register with their names.
+
+### Step 3: Implement feature-switch
+
+Once the registry is complete the installer core code must be changed. Provider-specific switch-case structures can be identified by searching for the usage of the `Name` constants in the `types` package. The existing switch-case should be wrapped in an if-else case. This will help with keeping the functionality of providers that have not yet switched.
+
+```go
+if v2, err := registry.Get(providerName); err == nil {
+    v2.CallSpecifiedFunction(...)
+} else if (!errors.Is(err, registry.ProviderNotFound)) {
+    return err
+} else {
+    switch (...) {
+        //...
+    }
+}
+```
+
+### Step 4: Migrate providers
+
+Providers need to be migrated to the new API.
```

Comment:
One of the biggest sources of friction we have had with "everyone needs to..." projects is prioritizing the work on the teams' backlogs. How many different teams will be involved in this part of the work? Are they all aware of this plan?

Reply:
@dhellmann I have sent out this plan to all teamleads and managers affected as far as I could tell. The benefit of this plan is that no team *needs to* do anything. In addition, I'm working on a way to make sure that API calls added in the future don't create an immediate requirement to do anything for other teams. PoC will still take a while (probably until new year).